peekingduck/training/src/data/dataset.py

```diff
@@ -0,0 +1,224 @@
+
+        self.list_IDs = df[x_col].values
+        self.targets = df[y_col].values if stage != "test" else None
+        self._on_epoch_end()
+
+    def __len__(self) -> int:
+        """Denotes the number of batches per epoch"""
+        return int(np.floor(len(self.list_IDs) / self.batch_size))
+
+    def __getitem__(self, index: int) -> Union[Tuple, Any]:
+        """Generate one batch of data
+
+        Args:
+            index (int): index of the image to return.
+
+        Returns:
+            (Dict):
+                Outputs dictionary with the keys `labels`.
+        """
+        # Generate indexes of the batch
+        indexes = self.indexes[index * self.batch_size : (index + 1) * self.batch_size]
+
+        # Find list of IDs
+        list_IDs_temp: List[Any] = [self.list_IDs[k] for k in indexes]
+
+        # Generate data
+        X, y = self._data_generation(list_IDs_temp, indexes)
+
+        if self.stage in ["train", "valid", "debug"]:
+            return X, y
+        elif self.stage == "test":
+            return X
+        else:
+            raise ValueError(f"Invalid stage {self.stage}.")
+
+    def load_image(self, image_path: str) -> Any:
+        """Load image from `image_path`
+
+        Args:
+            image_path (str): image path to load.
+
+        Returns:
+            (ndarray):
+                Outputs image in numpy array.
+        """
+        image = cv2.imread(image_path)  # BGR
+        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+
+        image = self.apply_image_transforms(image)
+        """ Preprocessed numpy.array or a tf.Tensor with type float32. The images are converted from RGB to BGR,
+            then each color channel is zero-centered with respect to the ImageNet dataset, without scaling. """
+        return image
+
+    def _data_generation(
+        self,
+        list_IDs_temp: List[Any],
+        indexes: np.ndarray,
+    ) -> Any:
+        "Generates data containing batch_size samples"  # X : (n_samples, *dim, n_channels)
+        # Initialization
+        X = np.empty((self.batch_size, *self.dim, self.num_channels))
+        y = np.empty((self.batch_size), dtype=int)
+
+        # Generate data
+        for i, ID in enumerate(list_IDs_temp):
+            # Store sample
+            X[i] = self.load_image(ID)
+
+        # Store class
+        y = self.targets[indexes]
+
+        return X, tf.keras.utils.to_categorical(y, num_classes=self.num_classes)
+
+    def apply_image_transforms(self, image: Any) -> Any:
+        """Apply transforms to the image."""
+        if self.transforms and isinstance(self.transforms, A.Compose):
+            image = self.transforms(image=image)["image"]
+        elif self.transforms and isinstance(self.transforms, T.Compose):
+            image = self.transforms(image)
+        return image
+
+    def _on_epoch_end(self) -> None:
+        "Updates indexes after each epoch"
+        self.indexes = np.arange(len(self.list_IDs))
+        if self.shuffle == True:
```

Comment:
Consider using `if self.shuffle` or `self.shuffle is True`. However, as the reference below mentioned, be mindful of using it with `numpy` and its derivatives.

See [reference on singleton-comparison](https://stackoverflow.com/questions/27276610/boolean-identity-true-vs-is-true).